src/pd/client.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright 2018 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+
+use futures::future::BoxFuture;
+use futures::future::{ready, Either};
+use futures::prelude::*;
+use futures::stream::BoxStream;
+use grpcio::EnvBuilder;
+
+use crate::{
+    compat::{stream_fn, ClientFutureExt},
+    kv::BoundRange,
+    kv_client::{KvClient, KvConnect, Store, TikvConnect},
+    pd::{Region, RegionId, RetryClient},
+    security::SecurityManager,
+    transaction::Timestamp,
+    Config, Key, Result,
+};
+
+const CQ_COUNT: usize = 1;
+const CLIENT_PREFIX: &str = "tikv-client";
+
+pub trait PdClient: Send + Sync + 'static {
+    type KvClient: KvClient + Send + Sync + 'static;
+
+    fn map_region_to_store(
+        self: Arc<Self>,
+        region: Region,
+    ) -> BoxFuture<'static, Result<Store<Self::KvClient>>>;
+
+    fn region_for_key(&self, key: &Key) -> BoxFuture<'static, Result<Region>>;
+    fn region_for_id(&self, id: RegionId) -> BoxFuture<'static, Result<Region>>;
+
+    fn store_for_key(
+        self: Arc<Self>,
+        key: &Key,
+    ) -> BoxFuture<'static, Result<Store<Self::KvClient>>> {
+        self.region_for_key(key)
+            .and_then(move |region| self.clone().map_region_to_store(region))
+            .boxed()
+    }
+
+    fn store_for_id(
+        self: Arc<Self>,
+        id: RegionId,
+    ) -> BoxFuture<'static, Result<Store<Self::KvClient>>> {
+        self.region_for_id(id)
+            .and_then(move |region| self.clone().map_region_to_store(region).boxed())
+            .boxed()
+    }
+
+    fn group_keys_by_region<K: AsRef<Key> + Send + Sync + 'static>(
+        self: Arc<Self>,
+        keys: impl Iterator<Item = K> + Send + Sync + 'static,
+    ) -> BoxStream<'static, Result<(RegionId, Vec<K>)>> {
```

Comment:
Stream may not be the best idea. I think we can get them concurrently.

Reply:
I don't understand what you mean exactly, streams can be handled concurrently

Reply:
I mean in a single request, if we have a stream of stores, then we will send requests in a stream. I expect that we can send all requests concurrently.
Am I misunderstanding how a `Stream` works?

Reply:
It depends exactly what you mean. We can execute requests concurrently using  [for_each_concurrent](https://docs.rs/futures-preview/0.3.0-alpha.17/futures/stream/trait.StreamExt.html#method.for_each_concurrent). That doesn't support using different threads, but I think that is an issue for `for_each_concurrent`, not streams in general. That means we can write our own Stream processor which uses threads (although I expect there will be one available soon, if not already).

Reply:
Cool. I didn't notice this function. I think it should satisfy our needs.